File: leetcode/code/75.sort-colors.py
class Solution:
    def sortColors(self, nums: list[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        i, i0, i2 = 0, 0, n - 1
        while i0 < i2 and i <= i2:
            match nums[i]:
                case 0:
                    nums[i], nums[i0] = nums[i0], nums[i]
                    i += 1
                    i0 += 1
                case 1:
                    i += 1
                case 2:
                    nums[i], nums[i2] = nums[i2], nums[i]
                    i2 -= 1
                case _:
                    raise ValueError

File: leetcode/code/75.sort-colors.rs
impl Solution {
    pub fn sort_colors(nums: &mut Vec<i32>) {
        let n = nums.len();
        let mut i0 = 0;
        let mut i2 = n - 1;
        let mut i = 0;
        while i < nums.len() && i0 < i2 && i <= i2 {
            match nums[i] {
                0 => {
                    nums.swap(i, i0);
                    i0 += 1;
                    i += 1;
                }
                1 => {
                    i += 1;
                }
                2 => {
                    nums.swap(i, i2);
                    i2 -= 1;
                }
                _ => unreachable!(),
            }
        }
    }
}
